src/storage/src/hummock/store/version.rs

```diff
@@ -780,31 +812,152 @@ impl HummockVersionReader {
         read_version_tuple: (Vec<ImmutableMemtable>, Vec<SstableInfo>, CommittedVersion),
         memtable_iter: MemTableHummockIterator<'a>,
     ) -> StorageResult<LocalHummockStorageIterator<'_>> {
+        let user_key_range_ref = bound_table_key_range(read_options.table_id, &table_key_range);
+        let user_key_range = (
+            user_key_range_ref.0.map(|key| key.cloned()),
+            user_key_range_ref.1.map(|key| key.cloned()),
+        );
+        let mut factory = ForwardIteratorFactory::default();
+        let mut local_stats = StoreLocalStatistic::default();
+        let (imms, uncommitted_ssts, committed) = read_version_tuple;
+        let table_id = read_options.table_id;
+        let min_epoch = gen_min_epoch(epoch, read_options.retention_seconds.as_ref());
         self.iter_inner(
             table_key_range,
             epoch,
             read_options,
-            read_version_tuple,
-            Some(memtable_iter),
+            imms,
+            uncommitted_ssts,
+            &committed,
+            &mut local_stats,
+            &mut factory,
         )
-        .await
+        .await?;
+        let merge_iter = factory.build(Some(memtable_iter));
+        // the epoch_range left bound for iterator read
+        let mut user_iter = UserIterator::new(
+            merge_iter,
+            user_key_range,
+            epoch,
+            min_epoch,
+            Some(committed),
+        );
+        user_iter.rewind().await?;
+        Ok(HummockStorageIteratorInner::new(
+            user_iter,
+            self.state_store_metrics.clone(),
+            table_id,
+            local_stats,
+        ))
     }
 
-    pub async fn iter_inner<'a, 'b>(
-        &'a self,
+    pub async fn rev_iter(
```

Comment:
It seems that `rev_iter` and `rev_iter_with_memtable` only differ in the `mem_table_iter` and `factory.build(...)`. 

Can we merge the two methods into a single method that takes `Option<MemTableHummockRevIterator<'a>,>` as input? Code may be like
```
pub async fn rev_iter_inner<'a, 'b>(
        &'a self,
        table_key_range: TableKeyRange,
        epoch: u64,
        read_options: ReadOptions,
        read_version_tuple: (Vec<ImmutableMemtable>, Vec<SstableInfo>, CommittedVersion),
        memtable_iter: Option<MemTableHummockRevIterator<'b>>,
    ) -> StorageResult<HummockStorageRevIteratorInner<'b>> {
    ...
    let merge_iter = factory.build(memtable_iter);
    ...
}
```

When `memtable_iter` is `None`, lifetime `'b` will become `'static`, and then the returned value will also be `'static`.

Ditto for the forward iter changed in this PR. The `iter` and `iter_with_memtable` can also be merged into a single method named like `forward_iter`.

Ideally, the four methods can be further merged to a single `iter_inner` as long as we provide a new method `async fn build<'a>(self, ..., Option<Self::MemtableIterType<'a>>) -> Self::UserIterType<'a>` method and define new associated types in the `IteratorFactory` trait, and within `iter_inner` we call `build` after collecting the iterators.

Reply:
Good suggestion